### PR TITLE
feat(revm): add forking mode

### PR DESCRIPTION
**Pull request by gakonst**
Ports over the forking mode that @mattsse wrote to REVM on top of @onbjerg's latest work.

Would appreciate some review from both.  @onbjerg can maybe integration test it with your test suite? @mattsse you can diff the `HEAD` against the first commit to see the changes I had to make, can you pls sanity check I didn't make some mistake, as this is quite a low level API handler?

### CODE REVIEW

File: forge/src/executor/fork/mod.rs [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+                            match listener {
+                                AccountListener::Basic(sender) => {
+                                    let _ = sender.send(acc.clone());
+                                }
+                            }
+                        }
+                        continue
+                    }
+                }
+                // TODO: How should we handle Storage now that it's on a separate place?
```
Comment (onbjerg) [inline forge/src/executor/fork/mod.rs:271]: Can you clarify?

**Review by onbjerg — COMMENTED** [review]
I'm assuming the intention is that this backend is wrapped in a `CacheDB` like we do now for `EmptyDB`. If so, LGTM overall. Chain ID and so on will be set outside of this database as a note

File: forge/src/executor/fork/mod.rs [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+    sync::{
+        mpsc::{channel as oneshot_channel, Sender as OneshotSender},
+        Arc,
+    },
+};
+
+use foundry_utils::RuntimeOrHandle;
+
+/// A basic in memory cache (address -> AccountInfo + Storage)
+pub type MemCache = BTreeMap<H160, (AccountInfo, BTreeMap<U256, U256>)>;
```
Comment (mattsse) [inline forge/src/executor/fork/mod.rs:28]: Perhaps worthwhile to make this a new struct, keeping `MemoryAccount`?

File: forge/src/executor/fork/mod.rs [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+        }
+    }
+
+    /// handle the request in queue in the future.
+    ///
+    /// We always check:
+    ///  1. if the requested value is already stored in the cache, then answer the sender
+    ///  2. otherwise, fetch it via the provider but check if a request for that value is already in
+    /// progress (e.g. another Sender just requested the same account)
+    fn on_request(&mut self, req: BackendRequest) {
```
Comment (mattsse) [inline forge/src/executor/fork/mod.rs:121]: this is the same (minus Exist/Code variants)

lgtm

File: forge/src/executor/fork/mod.rs [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+                entry.insert(vec![listener]);
+                let provider = self.provider.clone();
+                let block_id = self.block_id;
+                let fut = Box::pin(async move {
+                    // serialize & deserialize back to U256
+                    let idx_req = H256::zero(); // TODO, convert U256 to H256
+                    let _storage = provider.get_storage_at(address, idx_req, block_id).await;
+                    // TODO: convert H256 to U256
+                    let storage = Ok(U256::zero()); // TODO;
+                    (storage, address, idx)
```
Comment (mattsse) [inline forge/src/executor/fork/mod.rs:187]: so revm stores as U256, whereas provider returns H256?

File: forge/src/executor/fork/mod.rs [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+                    if let Poll::Ready((resp, addr)) = fut.poll_unpin(cx) {
+                        // get the response
+                        let (balance, nonce, code) = resp.unwrap_or_else(|_| {
+                            tracing::trace!("Failed to get account for {}", addr);
+                            Default::default()
+                        });
+
+                        // conver it to revm-style types
+                        let (code, code_hash) = if !code.0.is_empty() {
+                            (Some(code.0.clone()), keccak256(&code).into())
```
Comment (mattsse) [inline forge/src/executor/fork/mod.rs:250]: is this clone necessary here?

File: forge/src/executor/fork/mod.rs [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+
+                        // TODO: Is the storage stuff here still required?
+                        let (listeners, storage) =
+                            pin.account_requests.remove(&addr).unwrap_or_default();
+                        let acc = AccountInfo { nonce: nonce.as_u64(), balance, code, code_hash };
+                        pin.cache.write().insert(addr, (acc.clone(), storage));
+                        // notify all listeners
+                        for listener in listeners {
+                            match listener {
+                                AccountListener::Basic(sender) => {
```
Comment (mattsse) [inline forge/src/executor/fork/mod.rs:263]: ~~iirc `Basic` was nomenclature of sputnik, perhaps we rename that as well? `account(_info)`?~~
nvm there's still `Database::basic`

**Review by mattsse — COMMENTED** [review]
looks correct at this point.
perhaps we don't even need to take `db`? 
`InMemoryDB` appears to the same as the `MemCache` type, which also entangles storage from account.
I guess this would make the BackendHandler significantly easier, because there's no need to buffer the storage requests until the account was fetched anymore. with sputniks `MemoryAccount` that was the problem the reason this storage buffer was introduced was because `storage` is part of `sputnik::MemoryAccount`.

So I'm thinking we can perhaps get rid of the `the account not fetched yet, we either add this value to the storage` branch, if we use `type MemCache = InMemoryDB;`?

**Comment by gakonst** [comment]
I got fixes to all the comments and will push once I'm at computer!

File: forge/src/executor/fork/mod.rs [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+        self.do_get_storage(address, index).unwrap_or_else(|_| {
+            tracing::trace!("Failed to send/recv `storage` for {} at {}", address, index);
+            Default::default()
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+struct SharedBackendInner<DB> {
+    db: DB,
```
[the remark this answers is not in the document]
Comment (gakonst) [reply forge/src/executor/fork/mod.rs]: The DB was not needed at all, see latest state of the PR

File: forge/src/executor/fork/mod.rs (hunk 4 above)
Comment (onbjerg) [reply forge/src/executor/fork/mod.rs]: yep

File: forge/src/executor/builder.rs [hunk 8]
```diff
@@ -38,9 +41,25 @@ impl ExecutorBuilder {
...
+        let db: Box<dyn DatabaseRef> = if let Some(fork) = self.fork_url {
+            let provider = Provider::try_from(fork).unwrap();
+            // TODO: Add pin block
+            // TOOD: Add reading cache from disk
+            let backend = SharedBackend::new(provider, SharedMemCache::default(), None);
+            Box::new(backend)
+        } else {
+            Box::new(EmptyDB())
+        };
+        Executor::new(db, self.config)
```
Comment (onbjerg) [inline forge/src/executor/builder.rs:62]: wow this is considerably simpler than what we had for sputnik

**Comment by onbjerg** [comment]
LGTM overall with changes, will add a fork test to the test suite and get it checked

File: forge/src/executor/builder.rs (hunk 8 above)
Comment (gakonst) [reply forge/src/executor/builder.rs:62]: Still doesn't build - but almost there!

**Pull request merged**